Extend e3sm_prod testing with throughput and memory checks

amametjanov opened the pull request:
Extend e3sm_prod testing with throughput and memory checks:
* add XML variable `TEST_MEMLEAK_TOLERANCE` for Chrysalis
* add `e3sm_prod_bench` no-IO test suite

This helps detect throughput and memory regressions while integrating new features.

[BFB]

amametjanov commented:
This depends on ESMCI/cime#4151 .

amametjanov commented:
All throughput and memory regressions (while comparing with existing baselines) are currently ignored: i.e. tests pass even if there is a performance regression. This PR expects to run 5 production-like tests for 20 days without any netcdf output and compare SYPD and memory highwater at the end of the runs. An example of a throughput regression detection is at https://my.cdash.org/viewTest.php?buildid=2114051 .

sarats commented on cime_config/machines/config_machines.xml:
```diff
@@ -1503,7 +1503,8 @@
     <RUNDIR>$CIME_OUTPUT_ROOT/$CASE/run</RUNDIR>
     <EXEROOT>$CIME_OUTPUT_ROOT/$CASE/bld</EXEROOT>
-    <TEST_TPUT_TOLERANCE>0.1</TEST_TPUT_TOLERANCE>
+    <TEST_TPUT_TOLERANCE>0.05</TEST_TPUT_TOLERANCE>
```
So, this tolerance would account for system variability as well as changes in a specific PR. I presume 0.05 tolerance translates to 5%. 
Even after turning off output, there may be a few outliers where reads/communication might occasionally trigger a false positive. Hopefully we don't encounter too many of those.

rljacob commented:
Should a Jenkins service run e3sm_prod_bench on some schedule/machine ?

amametjanov commented:
Yes, it'd be good to start running [this test suite](https://github.com/E3SM-Project/E3SM_test_scripts/blob/master/jenkins/chrysalis_prod_bench.sh) on Chrysalis with the `next` branch daily. I'll submit a CDash request after the merge to `next`.

jgfouca commented:
@amametjanov , the latest CIME update is complete.

jgfouca commented:
You will need to do an upstream merge to your branch, resolve conflicts, and push.

rljacob commented:
Or just rebase.

rljacob commented:
What does "Rebase my cime branch from cime-master onto e3sm-cime" mean?  Are we no longer following cime master @jgfouca ?

jgfouca commented:
@rljacob , we still (almost always) follow CIME's master branch. I think @amametjanov may have misunderstood what I meant. I wanted him to either rebase or upstream merge his E3SM branch using E3SM's latest master (which contains the CIME subm update, which contains the CIME PR he wanted).

amametjanov commented:
Just noting the steps (to self). Straight rebase from e3sm root-dir didn't work.
```
@chrlogin1:E3SM-clone$ git rebase master
Fast-forwarding submodule cime
Auto-merging cime
error: add_cacheinfo failed to refresh for path 'cime'; merge aborting.
hint: Could not execute the todo command
hint: 
hint:     pick a981c175d64801eae9f720d9e5c2f7019d10a7e5 Update cime submodule with extended perf-checking branch
hint: 
hint: It has been rescheduled; To edit the command before continuing, please
hint: edit the todo list first:
hint: 
hint:     git rebase --edit-todo
hint:     git rebase --continue
Could not apply a981c175d6... Update cime submodule with extended perf-checking branch
@chrlogin1:E3SM-clone$ 
```
Had to go into `cime` and
```
@chrlogin1:cime$ git rebase --onto b0ada2143def2bfc5f6d42941486d8264a52628a master azamat/extend-perf-checks
Successfully rebased and updated refs/heads/azamat/extend-perf-checks.
@chrlogin1:cime$ 
```
which asks to rebase onto latest e3sm-cime commit b0ada2143 my branch, which branched from `cime`'s master.
Then go back to e3sm-root and rebase e3sm branch
```
@chrlogin1:E3SM-clone$ git rebase master
dropping a981c175d64801eae9f720d9e5c2f7019d10a7e5 Update cime submodule with extended perf-checking branch -- patch contents already upstream
dropping 7d8de3771d31e091b015785f3c67ae084fbba239 Rebase my cime branch from cime-master onto e3sm-cime -- patch contents already upstream
Successfully rebased and updated refs/heads/azamat/perf-tests/add-e3sm-prod-perf-tests.
@chrlogin1:E3SM-clone$ 
```
and force-push.

jgfouca commented:
@amametjanov , conflicts in submodules are a pain. I think there might have been an easier way to resolve it, but I can see that this PR no longer touches the CIME submodule, which is what we want. Merging to next.
